src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,519 @@
+            ulong intrinsicGas = CalculateIntrinsicGas(entry, spec);
+
+            if (_logger.IsEnabled(LogEventLevel.Verbose)) _logger.Verbose("Executing entry {entry}", entry);
+            if (!ValidateSender(entry, env, txTracer)) return;
+            if (!ValidateIntrinsicGas(entry, env, intrinsicGas, txTracer)) return;
+            if (!ValidateDeltaGasLimit(entry, env, txTracer)) return;
+
+            if (!_stateProvider.AccountExists(env.Sender))
+            {
+                if (env.GasPrice == UInt256.Zero) _stateProvider.CreateAccount(env.Sender, UInt256.Zero);
+            }
+
+            if (!ValidateSenderBalance(entry, env, intrinsicGas, txTracer)) return;
+            if (!ValidateNonce(entry, env, txTracer)) return;
+
+            InitEntryExecution(env, gasLimit, spec, txTracer);
+
+            // we prepare two fields to track the amount of gas spent / left
+            ulong unspentGas = gasLimit - intrinsicGas;
+            ulong spentGas = gasLimit;
+
+            // the snapshots are needed to revert the subroutine state changes in case of an VM exception
+            int stateSnapshot = _stateProvider.TakeSnapshot();
+            int storageSnapshot = _storageProvider.TakeSnapshot();
+
+            // we subtract value from sender
+            // it will be added to recipient at the later stage (inside the VM)
+            _stateProvider.SubtractFromBalance(sender, env.Value, spec);
+
+            // we fail unless we succeed
+            byte statusCode = StatusCode.Failure;
+            TransactionSubstate substate = null;
+
+            try
+            {
+                if (entry.IsValidDeploymentEntry) PrepareContractAccount(env.CodeSource);
+                ExecutionType executionType = entry.IsValidDeploymentEntry ? ExecutionType.Create : ExecutionType.Call;
+                using (VmState state = new VmState((long) unspentGas, env, executionType, isPrecompile, true, false))
+                {
+                    substate = _virtualMachine.Run(state, txTracer);
+                    unspentGas = (ulong) state.GasAvailable;
+                }
+
+                if (substate.ShouldRevert || substate.IsError)
+                {
+                    if (_logger.IsEnabled(LogEventLevel.Verbose)) _logger.Verbose("Restoring state from before transaction");
+                    _stateProvider.Restore(stateSnapshot);
+                    _storageProvider.Restore(storageSnapshot);
+                }
+                else
+                {
+                    if (entry.IsValidDeploymentEntry) DeployCode(env, substate, ref unspentGas, spec);
+                    DestroyAccounts(substate);
+                    statusCode = StatusCode.Success;
+                }
+
+                spentGas = Refund(gasLimit, unspentGas, substate, env, spec);
+            }
+            catch (Exception ex) when (ex is EvmException || ex is OverflowException) // TODO: OverflowException? still needed? hope not
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4350015)